velox/functions/sparksql/Hash.cpp

```diff
@@ -26,21 +26,284 @@ namespace {
 
 const int32_t kDefaultSeed = 42;
 
+// Computes the hash value of input using the hash function in HashClass.
+template <typename HashClass>
+typename HashClass::ReturnType hashOne(
+    int32_t input,
+    typename HashClass::SeedType seed) {
+  return HashClass::hashInt32(input, seed);
+}
+
+template <typename HashClass>
+typename HashClass::ReturnType hashOne(
+    int64_t input,
+    typename HashClass::SeedType seed) {
+  return HashClass::hashInt64(input, seed);
+}
+
+template <typename HashClass>
+typename HashClass::ReturnType hashOne(
+    float input,
+    typename HashClass::SeedType seed) {
+  return HashClass::hashFloat(input, seed);
+}
+
+template <typename HashClass>
+typename HashClass::ReturnType hashOne(
+    double input,
+    typename HashClass::SeedType seed) {
+  return HashClass::hashDouble(input, seed);
+}
+
+template <typename HashClass>
+typename HashClass::ReturnType hashOne(
+    int128_t input,
+    typename HashClass::SeedType seed) {
+  return HashClass::hashLongDecimal(input, seed);
+}
+
+template <typename HashClass>
+typename HashClass::ReturnType hashOne(
+    Timestamp input,
+    typename HashClass::SeedType seed) {
+  return HashClass::hashTimestamp(input, seed);
+}
+
+template <typename HashClass>
+typename HashClass::ReturnType hashOne(
+    StringView input,
+    typename HashClass::SeedType seed) {
+  return HashClass::hashBytes(input, seed);
+}
+
+template <typename HashClass, TypeKind kind>
+class PrimitiveVectorHasher;
+
+template <typename HashClass>
+class ArrayVectorHasher;
+
+template <typename HashClass>
+class MapVectorHasher;
+
+template <typename HashClass>
+class RowVectorHasher;
+
+// Class to compute hashes identical to one produced by Spark.
+// Hashes are computed using the algorithm implemented in HashClass.
+template <typename HashClass>
+class SparkVectorHasher {
+ public:
+  using SeedType = typename HashClass::SeedType;
+  using ReturnType = typename HashClass::ReturnType;
+
+  SparkVectorHasher(DecodedVector& decoded) : decoded_(decoded) {}
+
+  virtual ~SparkVectorHasher() = default;
+
+  // Compute the hash value of input vector at index.
+  ReturnType hashAt(vector_size_t index, SeedType seed) {
+    if (decoded_.isNullAt(index)) {
+      return seed;
+    }
+    return hashNotNullAt(index, seed);
+  }
+
+  // Compute the hash value of input vector at index for non-null values.
+  ReturnType hashNotNullAt(vector_size_t index, SeedType seed) {
+    switch (decoded_.base()->typeKind()) {
```

Comment:
Do we need to switch on every row? I wonder if it makes sense to do something similar to velox/functions/prestosql/aggregates/PrestoHasher.h

Reply:
I've looked into and tried `PrestoHasher` implementation but it cannot get the correct result for spark. In Spark's implementation https://github.com/apache/spark/blob/a2b7050e0fc5db6ac98db57309e4737acd26bf3a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L593-L607, the hash is computed for each row, and within the same row, the hash of one field is used as the seed for the next field. Therefore, it must switch with every row to produce identical results.

Reply:
@marin-ma I'm not suggesting using PrestoHasher as is, but rather implement Spark-specific hasher in a similar way.

> the hash is computed for each row, and within the same row, the hash of one field is used as the seed for the next field

I assume you refer to hashing values of type struct. If so, then it is the same for PrestoHasher. See `PrestoHasher::hash<TypeKind::ROW>`:

```
  for (int i = 0; i < baseRow->childrenSize(); i++) {
    children_[i]->hash(baseRow->childAt(i), elementRows, childHashes);
...
```

First, compute hashes for the first field for all rows. Then, use the results to compute hashes for the second field for all rows. Then third, etc.

Reply:
@mbasmanova If I understood the Spark behavior correctly, you can't do that as there is a data dependency between fully processing a row (and all its members if it's a struct/row) and processing the next row, as one hash output is used as the seed for the next hash. 

So changing the order to execute the column-first will change the result. This is bad design in the Spark side, but I'm not sure there is a good way to break this data dependency.

Reply:
@pedroerp Are you saying there is a dependency between rows? I don't think this makes sense as the order of rows being processed in not guaranteed in any way.

Reply:
> First, compute hashes for the first field for all rows. Then, use the results to compute hashes for the second field for all rows. Then third, etc.

@mbasmanova There are no dependency between input rows, but it cannot be computed that way to a specific row. Saying if a child field of the row type is type of `TypeKind::ROW`, i.e. the value in the row of that field is a struct, there is a dependency between the rows of that struct value. Therefore, the hash of this field should be computed row by row.

Reply:
@marin-ma I'm not sure I understand. Would you give an example?

Reply:
@mbasmanova Sorry that I've mixed up the struct and the array/map. There is a dependency from the order of the elements of array/map. https://github.com/apache/spark/blob/a2b7050e0fc5db6ac98db57309e4737acd26bf3a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L586-L590

Reply:
@mbasmanova 
```
  for (int i = 0; i < baseRow->childrenSize(); i++) {
    children_[i]->hash(baseRow->childAt(i), elementRows, childHashes);
...
```
Not sure if the implementation before commit https://github.com/facebookincubator/velox/pull/9414/commits/2ecc35c418ea31135a866512abb1137479b0c6b4 is more close to this design. Could you help to check it? The "switch" was introduced to avoid virtual function call in this commit.

Reply:
@marin-ma My concern is that type switching per value / row results in very poor performance. It would be nice to create a simple benchmark and compare performance before and after this change.

> There is a dependency from the order of the elements of array/map.

Got it. Still, it seems that it is possible to "vectorize" this computation. First, we compute hashes for all array elements. Then, we combine these hashes. This would be similar to Presto's logic. See `PrestoHasher::hash<TypeKind::ARRAY>` and would avoid type-switching per value.

Reply:
@mbasmanova 
>  First, we compute hashes for all array elements. Then, we combine these hashes.

We can not first compute hashes for all elements and then combine them, because the second element is computed using the hash of the first element as the seed. The hashes of the elements are not independent.

Reply:
@marin-ma My assumption is that hash function looks like this: hash(v, seed) = g(f(v), seed). Hence, hash of array is computed in a data-dependent loop over elements:

```
result = <initial value>;
for (auto e : elements) {
   result = hash(e, result);
}
```

This loop can be split into two loops:

```
elementHashes;
for (auto e : elements) {
   elementHashes.push_back(f(e));
}

result = <initial value>;
for (auto eh : elementHashes) {
   result = g(eh, result);
}
```

Now, the first loop is not data-dependent, only the second one is. Hence, we can process all elements at once and do not need to process each element separately.

Hope this makes sense.

Reply:
Example in below thread